src/main/scala/csv/CrimeRecord.scala

```diff
@@ -0,0 +1,126 @@
+
+/**
+  * Model for crime location
+  * @param longitude coordinates
+  * @param latitude coordinates
+  * @param location location
+  * @param crimetype type of crime in location
+  * @param locationCrimeNumber number of crime in location
+  */
+case class CrimeLocation(longitude: String, latitude: String, location: String, crimetype:String, locationCrimeNumber:String)
+
+/**
+  * Model for crime record.
+  * default value - ?
+  * @param crimeID ?
+  * @param month ?
+  * @param reportedby ?
+  * @param fallswithin ?
+  * @param longitude ?
+  * @param latitude ?
+  * @param location ?
+  * @param lSOAcode ?
+  * @param lSOAname ?
+  * @param crimetype ?
+  * @param lastoutcomecategory ?
+  * @param context ?
+  * @param locationCrimeNumber ?
+  */
+class CrimeRecord(var crimeID: String = "?", var month: String = "?",
+                  var reportedby: String = "?", var fallswithin: String = "?",
+                  var longitude: String = "?", var latitude: String = "?",
+                  var location: String = "?", var lSOAcode: String = "?",
+                  var lSOAname: String = "?", var crimetype: String = "?",
+                  var lastoutcomecategory: String = "?", var context: String = "?",
+                  var locationCrimeNumber: String = "?") {
+
+
+  /**
+    * Copy element into model
+    * @param position value position in model
+    * @param value value to insert into model position
+    * @return CrimeRecord
+    */
+  def copy(position: Int, value: String): CrimeRecord = {
+    var string = value.trim
+    if (string.isEmpty) string = "?"
+    position match {
+      case 13=>locationCrimeNumber=string
+      case 12 => crimeID = string
+      case 11 => month = string
+      case 10 => reportedby = string
+      case 9 => fallswithin = string
+      case 8 => longitude = string
+      case 7 => latitude = string
+      case 6 => location = string
+      case 5 => lSOAcode = string
+      case 4 => lSOAname = string
+      case 3 => crimetype = string
+      case 2 => lastoutcomecategory = string
+      case 1 => context = string
+    }
+    this
+  }
+
+  /**
+    * Model view in csv
+    * @return String
+    */
+  override def toString: String = s"${
+    crimeID
+  },${
+    month
+  },${
+    reportedby
+  },${
+    fallswithin
+  },${
+    longitude
+  },${
+    latitude
+  },${
+    location
+  },${
+    lSOAcode
+  },${
+    lSOAname
+  },${
+    crimetype
+  },${
+    lastoutcomecategory
+  },${
+    context
+  },${locationCrimeNumber}"
+}
+
+/**
+  * Record Builder
+  */
+object CrimeRecord {
+
+  /**
+    * Crime Record model builder
+    * @param list values to insert into model
+    * @return CrimeRecord
+    */
+  def build(list: List[String]): CrimeRecord = {
+    def go(i: Int, l: List[String]): CrimeRecord = {
+      l match {
+        case Nil => new CrimeRecord()
+        case head :: tail => go(i - 1, tail).copy(i, head)
+      }
+    }
+    go(12, list)
+  }
+
+}
```

Comment:
A simple case class should be used instead of that class. The case class with `Option`als fields will allow you to get rid of all that code, ie:

```scala
case class CrimeRecord(crimeID: Option[String], ...)

def parseCrimeRecord(csvLine: String): Option[CrimeRecord] = {
  val parts = csvLine.split(",")
  if (parts.size != expectedColumnsNumber) {
   None // invalid record
 } else {
   Some(CsvRecord(
    crimeId = parts.headOption,
    .... = parts(1)
  ))
 }
}
...
val csvFiles: List[String] = readAllFilesFromFolder() // lines from all files
val parsedRecords: List[CrimeRecord] = csvFiles.flatMap(parseCrimeRecord)

```

Reply:
Hi, sorry for trouble, from where you would get the expectedColumnsNumber

Reply:
@olehmv you have schema in the very first raw of the file.

Reply:
Something like this:
implicit class LToCSVRecord(l: List[String]) {
    def toCSVRecord:Option[CSVRecord] = {
        Some(CSVRecord(
          crimeId = Try(l.headOption).filter(op=> (op.get!="") & (op.get!="No Location")).getOrElse(None),
          month = Some(l(1)),
          reportedBy = Some(l(2)),
          fallsWithin = Some(l(3)),
          longitude = Some(l(4)),
          latitude = Some(l(5)),
          location = Some(l(6)),
          LSOACode = Some(l(7)),
          LSOAName = Some(l(8)),
          crimeType = Some(l(9)),
          lastOutcomeCategory = Try(Some(l(11))).getOrElse(None),
          context = Try(Some(l(11))).getOrElse(None)
        ))
      }
    }

  implicit object OrderByLocationCrimeNumber extends Ordering[((Option[String], Option[String]), List[CSVRecord])] {
    def compare(l1: ((Option[String], Option[String]), List[CSVRecord]), l2: ((Option[String], Option[String]), List[CSVRecord])) = {
      val l1Number = l1._2.size
      val l2Number = l2._2.size
      if (l1Number > l2Number) -1 else if (l1Number < l2Number) 1 else 0
    }
  }

 def res(dir:String): Unit ={
    new File(dir)
      .listFiles
      .filter(_.getName.endsWith(".csv"))
      .map(f => Source.fromFile(f).getLines().toList.drop(1)).toList
      .flatMap(l => l.map(s => s.split(",").toList))
      .flatMap(l => l.toCSVRecord)
      .filter(r => r.crimeId != None)
      .groupBy(r => (r.longitude, r.latitude))
      .toList.sorted
      .map(t=>((t._1._1.get,t._1._1.get)+":"+t._2.size+"\n",t._2.map(r=>r.crimeType.get).distinct.mkString("\n"))).take(5)
      .foreach(println)
  }

Reply:
Yes, but line's length from files vary from 10 to 12 whereas csv head is constant 12,
I used your advice with Option, and in place where value can be missed I wrap Option in Try